_posts/2023-01-06-Algorithms-Graphs-BFS-DFS.md

```diff
@@ -0,0 +1,338 @@
+
+#### BFS
+
+The visualization shows that we go level by level and I think this is the best way how we can imaging the algorithm .
+
+<p align="center">
+<img src="https://raw.githubusercontent.com/swiftingio/blog/%2357-Algorithms-DFS-BFS/BFS.svg" width="643">
+</p>
+
+If we start with node id = 3 (first layer), then we visit 1 and 4 (second layer) then 2 (third layer) and last will be 5 (fourth layer).
+
+### Impmenenation
+
+Having this knowledge let's implement DFS and BFS with Swift. Additionally let's take a look at some additional topics which are very interesting al well:
+
+- Finding the shortest path with unweighted graph
+- Finding cycle in an undirected graph
+- Finding cycle in a directed graph
+
+#### DFS
+
+```swift
+var visited: Set<Int> = []
+
+func dfs(_ root: TreeNode?)  {
+
+	guard let root = root else {
+		return
+	}
+
+	visited.insert(root.val)
+
+	for child in root.children {
+		if !visited.contains(child.val) {
+			dfs(child)
+		}
+	}
+}
+```
+
+#### BFS
+
+```swift
+func bfs(_ start: Node) {
+
+	var queue: [Node] = []
+	var visited: Set<Node> = []
+
+	queue.append(start)
+
+	while !queue.isEmpty { // maybe then this while is not needed anymore (worth to check)
+  	for item in queue {
+    	let first = queue.removeFirst()
+   		for child in first.children {
+				if !visited.contains(child) {
+      				visited.insert(child)
+						queue.append(child)
+					}
+				}
+			}
+		}
+	}	
+}
+```
+
+#### BFS + shortest path
+
+Implementation with additional finding distance from the start node and printing a path (an important note here is that here we are discussing graphs without weights. If we want to find the shortest distance for a graph with weights, we need to consider another algorithm which is Dijktstra).
+
+
+```swift
+func bfs(_ start: Node) {
+	var queue: [Node] = []
+	var visited: Set<Node> = []
+	var distance: [Node: Int] = [:]
+	var parent: [Node: Node] = [:]
+
+	queue.append(start)
+
+	while !queue.isEmpty {
+		for item in queue {
+			let first = queue.removeFirst()
+			for child in first.children {
+				if !visited.contains(child) {
+    	    		dist[child] = dist[item] + 1
+    	    		parent[child] = item 
+  					visited.insert(child)
+					queue.append(child)
+				}
+			}
+		}
+	}
+}
+
+func getPath(_ target: Node, _ start: Node) {
+  var x = target
+  var path: [Int] = []
+  
+  path.append(target)
+  
+  while x != start {
+    x = parent[x]
+    path.append(x)
+  }
+  
+  path.reverse()
+}
+```
+
+
+#### Detecting cycle in undirected graph
+
+```swift
+
+var visited: Set<Node> = []
+
+func dfs(_ root: TreeNode?) {	
+
+  guard let root = root else {	
+		return
+	}
+
+	visited.insert(root)
+	let children = root.children
+
+	for child in children {
+		if !visited.contains(child) {
+				dfs(child)
+		} else if child != previous {
+    		// here we can be sure that we don't have a cycle consists of two elements
```

Comment:
```suggestion
    		// here we can be sure that we don't have a cycle consisting of two elements
```

Reply:
Ta wersja DFS różni się od pierwszej wersji tylko tym `else` w którym nic się nie dzieje. Co powinno się zadziać w tej pętli ?

Reply:
Dobra uwaga, tutaj chodzi o to, że jeżeli znaleźliśmy cykl to upewnumay się, że to nie są dwa elementy leżące obok siebie. Dodam trochę więcej komentarzy i widzę, że zapomniałem dodać parametru stack